mlir/lib/Quantum/Transforms/emit_catalyst_pyface.cpp

```diff
@@ -0,0 +1,245 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "llvm/Support/FormatVariadic.h"
+
+#include "Quantum/Transforms/Passes.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using namespace mlir;
+using namespace catalyst::quantum;
+
+namespace {
+
+static size_t countCallsites(LLVM::LLVMFuncOp op)
+{
+    size_t count = 0;
+    op.walk([&](LLVM::CallOp callOp) { count++; });
+    return count;
+}
+
+static bool hasSingleCallsite(LLVM::LLVMFuncOp op) { return 1 == countCallsites(op); }
+
+static Optional<LLVM::LLVMFuncOp> getCallee(LLVM::LLVMFuncOp op)
+{
+    if (!hasSingleCallsite(op))
+        return std::nullopt;
+    Optional<LLVM::LLVMFuncOp> callee = std::nullopt;
+    op.walk([&](LLVM::CallOp callOp) {
+        auto calleeAttr = callOp.getCalleeAttr();
+        // calleeAttr is optional in case of function pointers.
+        if (!calleeAttr)
+            return;
+
+        callee = SymbolTable::lookupNearestSymbolFrom<LLVM::LLVMFuncOp>(op, calleeAttr);
+    });
+    return callee;
+}
```

Comment:
I think these three can be simplified and merged into a single function that only iterates once rather than twice

Reply:
I would rather keep these separate for legibility.

Reply:
Well my point was that the check for the number of callsites is unnecessary and a single iteration suffices:

Something like this should be just as legible:
```cpp
    Optional<LLVM::LLVMFuncOp> callee = std::nullopt;
    bool multipleCallees = false;

    op.walk([&](LLVM::CallOp callOp) {
        auto calleeAttr = callOp.getCalleeAttr();
        // calleeAttr is optional in case of function pointers.
        if (!calleeAttr)
            return;

        multipleCallees = callee.hasValue();
        callee = SymbolTable::lookupNearestSymbolFrom<LLVM::LLVMFuncOp>(op, calleeAttr);
    });

    return multipleCallees ? std::nullopt : callee;
```

Reply:
https://github.com/PennyLaneAI/catalyst/pull/106/commits/4c03d27ec01131312281860f5ad0c0a0026d127a